File: NeREF_training.py
import numpy as np
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, ReLU
from tensorflow.keras.optimizers import Adam
from sklearn.model_selection import train_test_split

# Load your data (assuming x and y are already in your workspace as NumPy arrays)
# For example:
# x = np.load('x.npy')
# y = np.load('y.npy')

# Check the shape of variables
print(x.shape)
print(y.shape)

# Split data into training and validation sets
x_train, x_val, y_train, y_val = train_test_split(x, y, test_size=0.2, random_state=42)

# Define the model architecture
NeREF = Sequential([
    Dense(128, input_shape=(x.shape[1],)),  # Input layer
    ReLU(),                                 # Activation function
    Dense(128),                             # Second hidden layer
    ReLU(),                                 # Activation function
    Dense(128),                             # Third hidden layer
    ReLU(),                                 # Activation function
    Dense(128),                             # Fourth hidden layer
    ReLU(),                                 # Activation function
    Dense(128),                             # Fifth hidden layer
    ReLU(),                                 # Activation function
    Dense(128),                             # Sixth hidden layer
    ReLU(),                                 # Activation function
    Dense(128),                             # Seventh hidden layer
    ReLU(),                                 # Activation function
    Dense(128),                             # Eighth hidden layer
    ReLU(),                                 # Activation function
    Dense(y.shape[1])                       # Output layer
])

# Compile the model
NeREF.compile(optimizer=Adam(learning_rate=0.001),
              loss='mean_squared_error')

# Train the model
history = NeREF.fit(x_train, y_train,
                    validation_data=(x_val, y_val),
                    epochs=50,
                    batch_size=64,
                    verbose=1)

# Save the model
NeREF.save('NeREF.h5')

File: NeREF_training.m
% Check the size of variables
disp(size(x))
disp(size(y))

% Split data into training and validation sets
numTrain = floor(0.8 * size(x, 1));
xTrain = x(1:numTrain, :);
yTrain = y(1:numTrain, :);
xVal = x(numTrain+1:end, :);
yVal = y(numTrain+1:end, :);

% Define the network architecture
layers = [
    imageInputLayer([size(x, 2) 1 1])  % Input layer
    fullyConnectedLayer(128)           % First hidden layer
    reluLayer                          % Activation function
    fullyConnectedLayer(128)           % Second hidden layer
    reluLayer                          % Activation function
    fullyConnectedLayer(128)           % Third hidden layer
    reluLayer                          % Activation function
    fullyConnectedLayer(128)           % Fourth hidden layer
    reluLayer                          % Activation function
    fullyConnectedLayer(128)           % Fifth hidden layer
    reluLayer                          % Activation function
    fullyConnectedLayer(128)           % Sixth hidden layer
    reluLayer                          % Activation function
    fullyConnectedLayer(128)           % Seventh hidden layer
    reluLayer                          % Activation function
    fullyConnectedLayer(128)           % Eighth hidden layer
    reluLayer                          % Activation function
    fullyConnectedLayer(size(y, 2))    % Output layer
    regressionLayer                    % Regression layer
];

% Training options
options = trainingOptions('adam', ... % Use Adam optimizer
    'MaxEpochs', 50, ...              % Maximum number of epochs
    'MiniBatchSize', 64, ...          % Mini-batch size
    'InitialLearnRate', 0.001, ...    % Initial learning rate
    'ValidationData', {reshape(xVal', [size(x, 2), 1, 1, size(xVal, 1)]), yVal}, ... % Validation data
    'Plots', 'training-progress', ... % Plot training progress
    'Verbose', false);                % Do not display verbose information

% Train the network
NeREF = trainNetwork(reshape(xTrain', [size(x, 2), 1, 1, size(xTrain, 1)]), yTrain, layers, options);

% Save the network
save('NeREF.mat', 'NeREF');
